cpp/src/arrow/compute/exec/vector_hash_test.cc

```diff
@@ -0,0 +1,234 @@
+// Licensed to the Apache Software Foundation (ASF) under one
+// or more contributor license agreements.  See the NOTICE file
+// distributed with this work for additional information
+// regarding copyright ownership.  The ASF licenses this file
+// to you under the Apache License, Version 2.0 (the
+// "License"); you may not use this file except in compliance
+// with the License.  You may obtain a copy of the License at
+//
+//   http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing,
+// software distributed under the License is distributed on an
+// "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+// KIND, either express or implied.  See the License for the
+// specific language governing permissions and limitations
+// under the License.
+
+#include <gmock/gmock-matchers.h>
+
+#include <chrono>
+#include <map>
+#include <set>
+#include "arrow/compute/exec/key_hash.h"
+#include "arrow/compute/exec/test_util.h"
+#include "arrow/compute/exec/util.h"
+#include "arrow/util/cpu_info.h"
+
+namespace arrow {
+namespace compute {
+
+void TestVectorHashImp(Random64Bit& random, bool use_32bit_hash, bool use_varlen_input,
```

Comment:
This test should be using `HashMultiColumn` as that is the public API and it isn't covered elsewhere and there is some non-trivial null handling logic there (we should also be testing with nulls).

In addition, there is no coverage of hash_bit

Finally, there is no coverage of multiple column hashes or hash combination

Reply:
Bit, null, HashMultiColumn are exercised by all hash join and hash group by tests, so the bugs in these areas would be caught there. Of course it is easier to debug failures in a smaller scenario, but bit, null, HashMultiColumn are build on top of core xhh hash arithmetic, so debugging from join or group by test is not unreasonably difficult.